lib/tasks/heroku.rake

```diff
@@ -0,0 +1,66 @@
+require 'git'
+namespace :wildland do
+  namespace :heroku do
+    desc 'Promotes staging to production.'
+    task promote_to_production: [:check_remotes, :check_heroku] do
+      turn_on_heroku_maintenance_mode
+      backup_production_database
+      promote_staging_to_production
+      migrate_production_database
+      turn_off_heroku_maintenance_mode
+    rescue RuntimeError => e
+      # Attempt to backout of the deploy
+    end
+
+    task :check_heroku do
+      unless heroku_toolbelt_available?
+        Kernal::abort('Missing heroku toolbelt')
+      end
+    end
+
+    task :check_remotes do
+      unless staging_remote_available?
+        Kernal::abort('Missing staging git remote')
+      end
+      unless production_remote_available?
+        Kernal::abort('Missing production git remote')
+      end
+    end
+  end
+end
+
+def turn_on_heroku_maintenance_mode
+  %w(staging production).each do |remote|
+    system("heroku maintenance:on -r #{remote}")
+  end
+end
+
+def turn_off_heroku_maintenance_mode
+  %w(staging production).each do |remote|
+    system("heroku maintenance:off -r #{remote}")
+  end
+end
+
+def backup_production_database
+  system('heroku pg:backups capture DATABASE -r production')
+end
+
+def promote_staging_to_production
+  system('heroku pipelines:promote -r staging')
+end
+
+def migrate_production_database
+  system('heroku run rake db:setup -r production')
+end
+
+def staging_remote_available?
+  Git.open('.').remotes.map(&:to_s).include?('staging')
+end
+
+def production_remote_available?
+  Git.open('.').remotes.map(&:to_s).include?('production')
+end
+
+def heroku_toolbelt_available?
+  system('which heroku > /dev/null 2>&1')
+end
```

Comment:
These should all be exposed as an API so that if a project needs to customize the heroku promotion process, they don't have to reimplement all of the stuff. Also feel that these would be better as rake tasks -- which would then also be public -- then it can just rely on each previous task to run.